pkg/consuldp/lifecycle.go

```diff
@@ -211,3 +216,60 @@ func (m *lifecycleConfig) gracefulShutdown() {
 	// Wait for context timeout to elapse
 	wg.Wait()
 }
+
+func (m *lifecycleConfig) gracefulStartupHandler(rw http.ResponseWriter, _ *http.Request) {
+
+	m.gracefulStartup()
+	//Unlike in gracefulShutdown, we want to delay the OK response until envoy is ready
+	//in order to block application container.
+	rw.WriteHeader(http.StatusOK)
+
+}
+
+// gracefulStartup blocks until the startup grace period has elapsed or we have confirmed that
+// Envoy proxy is ready.
+func (m *lifecycleConfig) gracefulStartup() {
+	m.logger.Info("Blocking container startup until Envoy ready / grace period elapsed")
+	timeout := time.Duration(m.shutdownGracePeriodSeconds) * time.Second
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
```

Comment:
**issue:** I am not 100% on this `loop:`. I think you may be able to send a message to ctx.Done() instead for both the timeout and ready aspect

Reply:
@trevorLeonHC I talk to my TL and this is what he suggested as maybe a better way to avoid the channel complications...

Something to consider...
```
func isReady() bool {
	time.Sleep(1 * time.Second)
	return true
}

func main() {
	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
	defer cancel()

	var ready atomic.Bool
	go func() {
		for ctx.Err() == nil {
			r := isReady()
			if r {
				ready.Store(true)
				cancel()
				break
			}
			// maybe put in a sleep of 50ms or something to prevent eating CPU if isReady could return quickly.
		}
	}()

	<-ctx.Done()
	fmt.Println("Got value", ready.Load())
}
```

Reply:
This was an awesome solution! I learned from it. Implemented!